---
output:
  BiocStyle::html_document
---


```{r setup, cache=FALSE, echo=FALSE, results='hide', message=FALSE}
library(knitr)

opts_chunk$set(cache=TRUE,
               cache.path="cache/QA",
               cache.extra=R.version.string,
               autodep=TRUE,
               fig.align="center",
               comment="")
```

# Data import

We start importing the raw table of counts from the SummarizedExperiment container. At this moment this contains only one assay, represented by a matrix-like object. The rows represent genes and the columns represent samples. 

```{r, message=FALSE}
library(SummarizedExperiment)

paired_se <- readRDS(file.path("results", "paired_seLUSC.rds"))
```

The row table of counts contains RNA-seq data from 20115 genes and 94 samples, which come from 47 patients. Samples are already paired.

We explore the column, that corresponds to samples, therefore it contains the phenotypic data, which in this case corresponds to clinical variables, and their corresponding metadata. 


```{r}
dim(colData(paired_se))
colData(paired_se)[1:5, 1:5]
mcols(colData(paired_se), use.names=TRUE)
```

These metadata consists of two columns of information about the clinical variables. One called `labelDescription` contains a succint description of the variable, often not more self-explanatory than the variable name itself, and the other called 'CDEID' corresponds to the `Common Data Element (CDE)` identifier. This identifier can be useed to search for further information about the associated clinical variable.

Now, we explore the row (feature) data.

```{r}
rowData(paired_se)
rowRanges(paired_se)
```

We are going to compare normal samples with tumor samples. Exploring the metadata, we can see that we have 47 normal samples and 47 tumor samples. As they are paired, the number must be the same.

```{r}
table(paired_se$type)
```

# Quality assessment

Normalitzation consists in the adjustment for sample and gene specific factors, to make gene expression values comparable across samples. This process is really important due to the fact that the samples may have been sequenced at different depth and that there may be sample specific biases related to technical differences in samples extarction, preparation and sequencing. 

The normalitzation is done in two steps: 
* Within-sample: adjustments to compare across features in a sample.
  *Scaling: using counts per million reads (CPM) mapped to the genome.
*Between-sample: adjustments to compare a feature across samples.
  *Sample-specific normalization factors: using the TMM algorithm from the R/Bioconductor package edgeR.
  *Quantile normalization: using the CQN algorithm from the R/Bioconductor package cqn or in the limma-voom pipeline.

To perform quality assessment and normalization we need first to load the [edgeR](http://bioconductor.org/packages/edgeR) R/Bioconductor package and create a `DGEList' object.

```{r, message=FALSE}
library(edgeR)

paired_dge <- DGEList(counts=assays(paired_se)$counts, genes=mcols(paired_se))
names(paired_dge)
saveRDS(paired_dge, file.path("results", "paired_dge.rds"))
```

Once the 'DGEList' object is created, we can performe the scaling to CPM values. Therefore, $\log_2$ CPM values of expression are calculated and used as an additional assay element to ease their manipulation. $\log_2$ CPM units separate better high and low expression, than raw counts or non-logged CPM units.

```{r}
assays(paired_se)$logCPM <- cpm(paired_dge, log=TRUE, prior.count=0.5)
assays(paired_se)$logCPM[1:5, 1:5]
```

## Sequencing depth

Let's examine the sequencing depth in terms of total number of sequence read counts
mapped to the genome per sample. Figure \@ref(fig:libsizes) below shows the
sequencing depth per sample, also known as library sizes, in increasing order.

<!---
you can control the height and width in pixels of the figure with 'out.height' and
'out.width'. Figures are automatically numbered, to refer to them in the main test
you should use the notation shown above as \@ref(fig:xxxx) with xxxx being the label
in the code chunk that also gives the filename of the figure. This name must be unique
--->

```{r libsizes, echo=FALSE, out.width="600px", fig.cap="Library sizes in increasing order."}
ord <- order(paired_dge$sample$lib.size/1e6)
barplot(paired_dge$sample$lib.size[ord], las=1, ylab="Millions of reads",
                xlab="Samples", col=c("blue", "red")[(paired_se$type[ord] == "tumor") + 1])
legend("topleft", c("normal", "tumor"), fill=c("red", "blue"), inset=0.01)
```

Now there is the same number of tumor samples and normal samples and they seem to be randonly distributed. However, there is still high variability in the library size.

## Distribution of expression levels among samples

Let's look at the distribution of expression values per sample in terms of
logarithmic CPM units. Due to the large number of samples, we display tumor
and normal samples separately, and are shown in Figure \@ref(fig:distRawExp)

```{r distRawExp, echo=FALSE, fig.height=4, fig.width=10, out.width="800px", fig.cap="Non-parametric density distribution of expression profiles per sample.", message=FALSE}
library(geneplotter)
par(mfrow=c(1, 2))
multidensity(as.list(as.data.frame(assays(paired_se[, paired_se$type == "tumor"])$logCPM)),
                          xlab="log 2 CPM", legend=NULL, main="Tumor samples", las=1)
multidensity(as.list(as.data.frame(assays(paired_se[, paired_se$type == "normal"])$logCPM)),
                          xlab="log 2 CPM", legend=NULL, main="Normal samples", las=1)
```

We do not appreciate substantial differences between the samples in the distribution of expression values.


## Distribution of expression levels among genes

Let's calculate now the average expression per gene through all the samples.
Figure \@ref(fig:exprdist) shows the distribution of those values across genes.

```{r exprdist, echo=FALSE, out.width="400px", fig.cap="Distribution of average expression level per gene."}
avgexp <- rowMeans(assays(paired_se)$logCPM)
hist(avgexp, xlab="log2 CPM", main="", las=1, col="steelblue")
abline(v=1, col="red", lwd=2)
```

RNA-seq expression profiles from lowly-expressed genes can lead to artifacts in downstream differential expression analyses. For this reason, it is common practice to remove them following some criteria, such as: filter out genes below a minimum average CPM (or log2 CPM) value throughout the samples or filter out genes with fewer than a given number of samples meeting a minimum CPM (or log2 CPM) cutoff. This graphic shows what would be the minimum average CPM (red line).


We can also make an MA-plot to see biases due to expression. First, we define the groups that we want to compare. In our case, we use the sample type to define groups, by modifying the DGEList object as follows:

```{r}
paired_dge$samples$group <- paired_se$type
table(paired_dge$samples$group)
```

Here we generate the MA-plot using type sample as grouping factor. 

```{r smear1, echo=FALSE, out.width="400px", fig.cap="MA-plot using type grouping."}
plotSmear(paired_dge, lowess = TRUE, las = 1, cex.lab = 1.5, cex.axis = 1.2)
abline(h = 0, col = "blue", lwd = 2)
```

Figure \@ref(fig:smear1) shows us the need of remove the lower experssed genes to normalize the samples.


## Filtering of lowly-expressed genes

In the light of this plot, we may consider a cutoff of 1 log CPM unit as minimum value
of expression to select genes being expressed across samples. Using this cutoff we proceed
to filter out lowly-expressed genes.

```{r}
mask <- avgexp > 1
dim(paired_se)
paired_se.filt <- paired_se[mask, ]
dim(paired_se.filt)
paired_dge.filt <- paired_dge[mask, ]
dim(paired_dge.filt)
```

Store un-normalized versions of the filtered expression data.

```{r}
saveRDS(paired_se.filt, file.path("results", "paired_se.filt.unnorm.rds"))
saveRDS(paired_dge.filt, file.path("results", "paired_dge.filt.unnorm.rds"))
```

We can also use a second approach to filter data using the CPM cutoff value of expression.
We will keep only genes that have this minimum level of expression in at least as many samples as the smallest group of comparison. We are still comparing sample type. 

```{r}
cpmcutoff <- round(10/min(paired_dge$sample$lib.size/1e+06), digits = 1)
cpmcutoff
nsamplescutoff <- min(table(paired_se$type))
nsamplescutoff
```

After knowing these parameters we can proceed to mask the lower-expressed genes. 

```{r}
mask2 <- rowSums(cpm(paired_dge) > cpmcutoff) >= nsamplescutoff
paired_se.filt2 <- paired_se[mask2, ]
paired_dge.filt2 <- paired_dge[mask2, ]
dim(paired_se)
dim(paired_se.filt)
dim(paired_se.filt2)
```

Store un-normalized versions of the filtered expression data.

```{r}
saveRDS(paired_se.filt2, file.path("results", "paired_se.filt2.unnorm.rds"))
saveRDS(paired_dge.filt2, file.path("results", "paired_dge.filt2.unnorm.rds"))
```

We will compare both approaches done before in order to compare them and see which one discards more genes and is more restrictive.

```{r exprdist2, echo=FALSE, out.width="400px", fig.cap="Distribution of average expression level per gene and filtering comparative."}
par(mar = c(4, 5, 1, 1))
h <- hist(avgexp, xlab="log2 CPM", main="", 
    las=1, col ="steelblue", cex.axis=1.2, cex.lab=1.5)
x <- cut(rowMeans(assays(paired_se.filt)$logCPM), breaks=h$breaks)
lines(h$mids-0.1, table(x), type = "h", lwd=5, lend=1, col = "olivedrab")
y <- cut(rowMeans(assays(paired_se.filt2)$logCPM), breaks=h$breaks)
lines(h$mids+0.1, table(y), type = "h", lwd=5, lend=1, col = "lightgoldenrod")
legend("topright", c("All genes", "Filtered genes", "Filtered 2 genes"), fill=c("steelblue", "olivedrab", "lightgoldenrod"))
```

After comparing the different approaches used in the filter of low expressed genes, we have decided to contunue working with the first set of filtered genes because this is more restrictive. We can see in Figure \@ref(fig:exprdist2) that the second approach is more conservative with our dataset. 

Therefore, we will make a first attempt using the first approach, which is more restrictive. If the accuracy and/or the sample size is not enough during the workflow, then we will consider to change the method of discarding genes.

## Normalization

We calculate now the normalization factors on the filtered expression data set.

```{r}
paired_dge.filt <- calcNormFactors(paired_dge.filt)
```

Replace the raw log2 CPM units in the corresponding assay element of the `SummarizedExperiment`
object, by the normalized ones.

```{r}
assays(paired_se.filt)$logCPM <- cpm(paired_dge.filt, log=TRUE, normalized.lib.sizes=TRUE, prior.count=0.25)
```

Store normalized versions of the filtered expression data.

```{r}
saveRDS(paired_se.filt, file.path("results", "paired_se.filt.rds"))
saveRDS(paired_dge.filt, file.path("results", "paired_dge.filt.rds"))
```


## MA-plots

We examine now the MA-plots of the normalized expression profiles. We look first to
the tumor samples in Figure \@ref(fig:maPlotsTumor).

<!---
Here we make a MA-plot for each sample. The options 'fig.height' and 'fig.width'
control the relative image size in *inches*. The final image size results from
'height'x'dpi' and 'width'x'dpi', where 'dpi' is the image resolution in
"dots per inch" (by default dpi=72). To scale the image to a desired size use
'out.width' and 'out.height'. More information at http://yihui.name/knitr/options
--->

First, we define the groups that we want to compare. In our case, we use the sample type to define groups, by modifying the DGEList object as follows:

```{r}
paired_dge$samples$group <- paired_se$type
table(paired_dge$samples$group)
paired_dge.filt$samples$group <- paired_se$type
```

Here we generate the MA-plot using type sample as grouping factor. 

```{r smear2, echo=FALSE, out.width="400px", fig.cap="MA-plot using type grouping after filtering by low expression."}
plotSmear(paired_dge.filt, lowess = TRUE, las = 1, cex.lab = 1.5, cex.axis = 1.2)
abline(h = 0, col = "blue", lwd = 2)
```

In Figure \@ref(fig:smear2) we don't see the artifact caused by the discreteness of counts at low values, where ratios between low numbers may easy lead to large fold-changes. In general, fold-changes from large expression values are more reliable than those coming from low-expression values. We don't see it because we have filtered by low-expression previously. But we see a majority of genes that seem to be upregulated under conditions where we do not expect it. This is because the proportion of reads attributed to a given gene in a sample depends on the expression properties of the whole sample, rather than just the expression level of that gene.

```{r maPlotsTumor, fig.height=36, fig.width=6, dpi=100, echo=FALSE, fig.cap="MA-plots of the tumor samples."}
par(mfrow=c(22, 3), mar=c(4, 5, 3, 1))
setmp <- paired_se.filt[, paired_se.filt$type == "tumor"]
dgetmp <- paired_dge.filt[, paired_se.filt$type == "tumor"]
for (i in 1:ncol(setmp)) {
  A <- rowMeans(assays(setmp)$logCPM)
  M <- assays(setmp)$logCPM[, i] - A
  samplename <- substr(as.character(setmp$bcr_patient_barcode[i]), 1, 12)
  smoothScatter(A, M, main=samplename, las=1)
  abline(h=0, col="blue", lwd=2)
  lo <- lowess(M ~ A)
  lines(lo$x, lo$y, col="red", lwd=2)
}
```

We do not observe samples with major expression-level dependent biases. Let's
look now to the normal samples in Figure \@ref(fig:maPlotsTumor).

```{r maPlotsNormal, fig.height=18, fig.width=6, dpi=100, echo=FALSE, fig.cap="MA-plots of the normal samples."}
par(mfrow=c(9, 3), mar=c(4, 5, 3, 1))
setmp <- paired_se.filt[, paired_se.filt$type == "normal"]
dgetmp <- paired_dge.filt[, paired_se$type == "normal"]
for (i in 1:ncol(setmp)) {
  A <- rowMeans(assays(setmp)$logCPM)
  M <- assays(setmp)$logCPM[, i] - A
  samplename <- substr(as.character(setmp$bcr_patient_barcode[i]), 1, 12)
  smoothScatter(A, M, main=samplename, las=1)
  abline(h=0, col="blue", lwd=2)
  lo <- lowess(M ~ A)
  lines(lo$x, lo$y, col="red", lwd=2)
}
```

We do not observe neither important expression-level dependent biases among the normal samples in Figure \@ref(fig:maPlotsNormal).

## Batch identification

We will search now for potential surrogate of batch effect indicators. Given that each sample
names corresponds to a TCGA barcode (see https://wiki.nci.nih.gov/display/TCGA/TCGA+barcode),
following the strategy described in http://bioinformatics.mdanderson.org/main/TCGABatchEffects:Overview
we are going to derive different elements of the TCGA barcode and examine their distribution
across samples.

```{r}
tss <- substr(colnames(paired_se.filt), 6, 7)
table(tss)
center <- substr(colnames(paired_se.filt), 27, 28)
table(center)
plate <- substr(colnames(paired_se.filt), 22, 25)
table(plate)
portionanalyte <- substr(colnames(paired_se.filt), 18, 20)
table(portionanalyte)
samplevial <- substr(colnames(paired_se.filt), 14, 16)
table(samplevial)
```

From this information we can make the following observations:

  * All samples were sequenced at the same center

  * Samples belong to three vials.

  * Samples were collected across different tissue source sites (TSS).

  * Samples were sequenced within differnt plates. There are 26 different plates.

  * Samples were sequenced using different portions and analyte combinations. There are 10 different conditions described. 
  

We are going to use the TSS as surrogate of batch effect indicator. Considering our outcome
of interest as molecular changes between sample types, tumor vs. normal, we will examine now
the cross-classification of this outcome with TSS.

```{r}
table(data.frame(TYPE=paired_se.filt$type, TSS=tss))
```

Observe that normal tissues of many TSS are under-represented with respect to
the tumor tissues. If TSS is a source of expression variability, this under-representation
of those TSS in the normal samples may lead to a potential confounding effect.

We can also examine the other three parameters that can lead to variability due to technical issues. We are not interested in this variability and can be a source of confussion.

```{r}
table(data.frame(TYPE=paired_se.filt$type, PORTIONALYTE=portionanalyte))
table(data.frame(TYPE=paired_se.filt$type, PLATE=plate))
table(data.frame(TYPE=paired_se.filt$type, SAMPLEVIAL=samplevial))
```

* All normal samples belong to the same vial.  

* Normal samples only have two different kind of portion and analyte combinations. 

We examine now how samples group together by hierarchical clustering and multidimensional
scaling, annotating the outcome of interest and the the surrogate of batch indicator. We
calculate again log CPM values with a higher prior count to moderate extreme fold-changes
produced by low counts. We start analysing the tss factor. The resulting dendrogram is shown in Figure \@ref(fig:sampleClustering1).

```{r sampleClustering1, fig.height=7, fig.width=14, dpi=100, echo=TRUE, fig.cap="Hierarchical clustering of the samples."}
logCPM <- cpm(paired_dge.filt, log=TRUE, prior.count=3)
d <- as.dist(1-cor(logCPM, method="spearman"))
sampleClustering <- hclust(d)
batch <- as.integer(factor(tss))
sampleDendrogram <- as.dendrogram(sampleClustering, hang=0.1)
names(batch) <- colnames(paired_se.filt)
outcome <- paste(substr(colnames(paired_se.filt), 9, 12), as.character(paired_se.filt$type), sep="-")
names(outcome) <- colnames(paired_se.filt)
sampleDendrogram <- dendrapply(sampleDendrogram,
                               function(x, batch, labels) {
                                 if (is.leaf(x)) {
                                   attr(x, "nodePar") <- list(lab.col=as.vector(batch[attr(x, "label")]))
                                   attr(x, "label") <- as.vector(labels[attr(x, "label")])
                                 }
                                 x
                               }, batch, outcome)
plot(sampleDendrogram, main="Hierarchical clustering of samples")
legend("topright", paste("Batch", sort(unique(batch)), levels(factor(tss))), fill=sort(unique(batch)))
```

We can not observe clear samples cluster. TSS  doesn't seem to have a stronger effect among a preferent kind of sample. For this reason we don't need to consider discarding samples leading to an unbalanced distribution of the outcome across batches.

In Figure \@ref(fig:mdsPlot1) we show the corresponding MDS plot. As the previous one this plot doesn't disclose any important batch effect related to TSS.

```{r mdsPlot1, fig.height=7, fig.width=14, dpi=100, echo=TRUE, fig.cap="Multidimensional scaling plot of the samples."}
plotMDS(paired_dge.filt, labels=outcome, col=batch)
legend("bottomleft", paste("Batch", sort(unique(batch)), levels(factor(tss))),
       fill=sort(unique(batch)), inset=0.05)
```

Now that we know that there is any batch effect of the tss, we are going to see what happen with the vial sample. We have decided to analyse this factor because the tables show us all normal samples belong to one vial since tumor samples belong to two other vials. This can probably give a batch effect. This means that the experiment is not balanced by this batch (sample vial) because individuals from every outcome do not occur through every batch. 

```{r sampleClustering2, fig.height=7, fig.width=14, dpi=100, echo=TRUE, fig.cap="Hierarchical clustering of the samples."}
logCPM <- cpm(paired_dge.filt, log=TRUE, prior.count=3)
d <- as.dist(1-cor(logCPM, method="spearman"))
sampleClustering <- hclust(d)
batch <- as.integer(factor(samplevial))
sampleDendrogram <- as.dendrogram(sampleClustering, hang=0.1)
names(batch) <- colnames(paired_se.filt)
outcome <- paste(substr(colnames(paired_se.filt), 9, 12), as.character(paired_se.filt$type), sep="-")
names(outcome) <- colnames(paired_se.filt)
sampleDendrogram <- dendrapply(sampleDendrogram,
                               function(x, batch, labels) {
                                 if (is.leaf(x)) {
                                   attr(x, "nodePar") <- list(lab.col=as.vector(batch[attr(x, "label")]))
                                   attr(x, "label") <- as.vector(labels[attr(x, "label")])
                                 }
                                 x
                               }, batch, outcome)
plot(sampleDendrogram, main="Hierarchical clustering of samples")
legend("topright", paste("Batch", sort(unique(batch)), levels(factor(samplevial))), fill=sort(unique(batch)))
```

We can observe that samples cluster primarily by sample type: tumor or normal. Sample vial seems to have a stronger effect among the normal and tumor samples. We may consider discarding samples leading to an unbalanced distribution of the outcome across batches, because this creates a correlation between outcome and batch that can potentially mask true biological differences and lead to incorrect conclusions.

In \@ref(fig:mdsPlot2) we show the corresponding MDS plot. Here we see more clearly that this source of variation separates tumor from normal samples. We can also observe that two tumor samples, corresponding to individuals 4129 and 6560 are separated from the rest, just as it happens in the hierchical clustering. A closer examination of their corresponding MA-plots also reveals a slight dependence of expression changes on average expression. We may consider discarding these two samples and doing the MDS plot again to have a closer look to the differences among the rest of the samples and their relationship with TSS.


```{r mdsPlot2, fig.height=7, fig.width=14, dpi=100, echo=TRUE, fig.cap="Multidimensional scaling plot of the samples."}
plotMDS(paired_dge.filt, labels=outcome, col=batch)
legend("bottomleft", paste("Batch", sort(unique(batch)), levels(factor(samplevial))),
       fill=sort(unique(batch)), inset=0.05)
```

For the former case, assume that from the previous MDS plot we could decide discard samples of !!!!!!!!!!sample vial 4129 and 6560????. But we will not do it because we have problems with the matrix size after removing this two samples. 

We will remove the batch effect that is given by the sample vial. We start building the design matrix of the full model Next, create the design matrix of the null model where, in this case, we are going to adjust for concentration. When there were no factor to adjust we would just give the intercept term by writing ~ 1 in the model formula:

```{r}
mod <- model.matrix(~type, data = colData(paired_se.filt))
head(mod)
```

To remove the batch effect observed (vial samples) we start using QR decomposition:

```{r}
library(limma)
qrexp <- removeBatchEffect(logCPM, batch, design = mod)
class(qrexp)
dim(qrexp)
```

We examine how the sample cluster looks like now:

```{r}
d <- as.dist(1 - cor(qrexp, method = "spearman"))
sampleClustering <- hclust(d)
sampleDendrogram <- as.dendrogram(sampleClustering, hang = 0.1)
names(batch) <- colnames(paired_se.filt)
outcome <- as.character(paired_se.filt$type)
names(outcome) <- colnames(paired_se.filt)
sampleDendrogram <- dendrapply(sampleDendrogram, function(x, batch, labels) {
    ## for every node in the dendrogram if it is a leaf node
    if (is.leaf(x)) {
        attr(x, "nodePar") <- list(lab.col = as.vector(batch[attr(x, "label")]))  ## color by batch
        attr(x, "label") <- as.vector(labels[attr(x, "label")])  ## label by outcome
    }
    x
}, batch, outcome)  ## these are the second and third arguments in the function
```

```{r dendogram, fig.height=7, fig.width=14, dpi=100, echo=TRUE, fig.cap="Multidimensional scaling plot of the samples."}
plot(sampleDendrogram, main = "Hierarchical clustering of samples")
legend("topright", paste("Batch", sort(unique(batch))), fill = sort(unique(batch)))
```



#The batch effect seems to be still there. For this reason, we have decided to try a more agressive approach to remove the batch effect.


#### JO NO FICARIA AIXÒ, PERQUÈ ABANS LES MOSTRES ESTAN SEPARADES PER TUMOR/NORMAL I DESPRÉS NO. A MÉS, AQUEST MÈTODE ASSUMEIX QUE LA MAJOR VARIABILITAT ENTRE MOSTRES VE DE BATCH EFFECTS, I POT SER NO ES EL NOSTRE CAS.
```{r}
library(corpcor)
s <- fast.svd(t(scale(t(logCPM), center = TRUE, scale = TRUE)))
pcSds <- s$d
pcSds[1] <- 0
```

```{r}
svdexp <- s$u %*% diag(pcSds) %*% t(s$v)
colnames(svdexp) <- colnames(paired_se.filt)
class(svdexp)
dim(svdexp)
```

We examine the cluster again:

```{r}
d <- as.dist(1 - cor(svdexp, method = "spearman"))
sampleClustering <- hclust(d)
sampleDendrogram <- as.dendrogram(sampleClustering, hang = 0.1)
names(batch) <- colnames(paired_se.filt)
outcome <- as.character(paired_se.filt$type)
names(outcome) <- colnames(paired_se.filt)
sampleDendrogram <- dendrapply(sampleDendrogram, function(x, batch, labels) {
    ## for every node in the dendrogram if it is a leaf node
    if (is.leaf(x)) {
        attr(x, "nodePar") <- list(lab.col = as.vector(batch[attr(x, "label")]))  ## color by batch
        attr(x, "label") <- as.vector(labels[attr(x, "label")])  ## label by outcome
    }
    x
}, batch, outcome)  ## these are the second and third arguments in the function
```

```{r}
plot(sampleDendrogram, main = "Hierarchical clustering of samples")
legend("topright", paste("Batch", sort(unique(batch))), fill = sort(unique(batch)))
```

Finally, the batch has disapear. 

#### SAVE 

We save the adjusted data: ## QUAN HAS MODIFICAT EL se.fils PERQUÈ ESTIGUI SENSE ELS BATCHES?

```{r}
saveRDS(paired_se.filt, file.path("results", "paired_se.filt.batch.rds"))
saveRDS(paired_dge.filt, file.path("results", "paired_dge.filt.batch.rds"))
```

## Session information

```{r, message=FALSE}
sessionInfo()
```


